framework/src/utils/SolutionInvalidity.C

```diff
@@ -0,0 +1,182 @@
+//* This file is part of the MOOSE framework
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "SolutionInvalidity.h"
+
+// MOOSE Includes
+#include "MooseError.h"
+#include "MooseApp.h"
+#include "VariadicTable.h"
+
+// System Includes
+#include <chrono>
+#include <memory>
+#include <timpi/parallel_sync.h>
+
+// libMesh Includes
+#include "libmesh/parallel_algebra.h"
+#include "libmesh/parallel_sync.h"
+
+SolutionInvalidity::SolutionInvalidity(MooseApp & app)
+  : ConsoleStreamInterface(app),
+    ParallelObject(app.comm()),
+    _solution_invalidity_registry(moose::internal::getSolutionInvalidityRegistry())
+{
+}
+
+void
+SolutionInvalidity::flagInvalidSolutionInternal(InvalidSolutionID _invalid_solution_id)
+{
+  std::lock_guard<std::mutex> lock_id(_invalid_mutex);
+  if (_counts.size() <= _invalid_solution_id)
+    _counts.resize(_invalid_solution_id + 1);
+
+  ++_counts[_invalid_solution_id].counts;
+}
+
+bool
+SolutionInvalidity::solutionInvalid() const
+{
+  libmesh_parallel_only(comm());
+
+  bool is_invalid = false;
+  for (auto & entry : _counts)
+  {
+    if (entry.counts)
+    {
+      is_invalid = true;
+      break;
+    }
+  }
+
+  comm().max(is_invalid);
+  return is_invalid > 0;
+}
+
+void
+SolutionInvalidity::resetSolutionInvalid()
+{
+  for (auto & entry : _counts)
+    entry.counts = 0;
+}
+
+void
+SolutionInvalidity::resetSolutionInvalidTimeIter()
+{
+  for (auto & entry : _counts)
+    entry.timeiter_counts = 0;
```

Comment:
It looks like this is never hit - can we come up with a lab case?

Reply:
I agree, I'll add a transient test for that.